### PR TITLE
Use data type from extension instead of generating json

### PR DESCRIPTION
**Pull request by thijsc**

### CODE REVIEW

File: lib/appsignal/utils.rb [hunk 1]
```diff
@@ -49,8 +118,7 @@ def encode_utf8(value)
           )
         end
       end
-
-      extend ClassMethods
```
Comment (tombruijn) [inline lib/appsignal/utils.rb]: This has already been changed in #169

Comment (thijsc) [inline lib/appsignal/utils.rb]: Ok, we'll merge that one first.

File: lib/appsignal/utils.rb [hunk 2]
```diff
@@ -14,12 +14,81 @@ def sanitize(params, only_top_level = false, key_sanitizer = nil)
     end
     extend ClassMethods
 
+    def self.data_generate(body)
+      Data.generate(body)
+    end
```
Comment (tombruijn) [inline lib/appsignal/utils.rb]: I don't think we need to add convenience methods on Appsignal::Utils for every util class we add

Comment (thijsc) [inline lib/appsignal/utils.rb]: Agree in general. I mirrored the existing Json implementation to make the change as small as possible.

File: spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb [hunk 3]
```diff
@@ -81,7 +81,7 @@
         transaction.should receive(:finish_event).with(
           'query.mongodb',
           'find | test | SUCCEEDED',
-          "{\"foo\":\"?\"}",
+          instance_of(Appsignal::Extension::Data),
```
Comment (tombruijn) [inline spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb]: Are we not using a lot of useful testing here by only by only testing the type?

Can we create an instance of Appsignal::Extension::Data with these values instead?

Comment (thijsc) [inline spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb]: That's possible, but quite a bit of work. Not sure that the added value is there since we have separate tests for the sanitizer that's used.

Comment (tombruijn) [inline spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb]: I see we have specs that test the generator, but wouldn't be good to see if the data we expect in the `Data` object to be what we expect?

Can't really create a solution for it since I can't install the extension. See also the travis build

File: spec/lib/appsignal/utils_spec.rb [hunk 4]
```diff
@@ -1,6 +1,59 @@
...
+  describe ".data_generate" do
+    subject { Appsignal::Utils.data_generate(body).to_s }
+
+    context "with a valid hash body" do
+      let(:body) do
+        {
+          'the' => 'payload',
+          1 => true,
+          nil => 'test',
+          :foo => [1, 2, 'three'],
```
Comment (matsimitsu) [inline spec/lib/appsignal/utils_spec.rb]: Maybe add a hash or even a child array here? just to cover _everything_ :)
```
:foo => [1, 2, 'three', [4,5], {'six' => 'seven'}]
```

File: lib/appsignal/utils.rb [hunk 5]
```diff
@@ -14,12 +14,81 @@ def sanitize(params, only_top_level = false, key_sanitizer = nil)
...
+            when Array
+              array.append_data(map_array(value))
+            else
+              array.append_string(value.to_s)
+            end
+          end
+          array
+        end
+      end
+    end
```
Comment (tombruijn) [inline lib/appsignal/utils.rb]: Can we move this to a separate file in `lib/appsignal/utils/data.rb`
I also want to move the JSON one to that dir eventually

And what kind of data does it generate? The name doesn't really convey that.

File: lib/appsignal/utils.rb (hunk 1 above)
Comment (tombruijn) [inline lib/appsignal/utils.rb]: was this fixed already? because you can undo this change. How it is now, changing it from a module to a singleton class, makes this a completely type of different object

File: spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb (hunk 3 above)
Comment (tombruijn) [inline spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb]: Aha, even if we create exactly the same data object it's comparison is done on pointer level and not values. It's probably not easy to make a comparison method for the internal values.

Can we move the conversion to the `Data` object later in the process? When it's received by `finish_event` for example

Or maybe even better: wrap it in an event formatter that converts all hashes and arrays? Rather than let individual integrations deal with the specifics of creating a valid body

File: lib/appsignal/utils.rb (hunk 1 above)
Comment (thijsc) [inline lib/appsignal/utils.rb]: I'm going to rebase this pull once I merge #134 and make this change.

File: lib/appsignal/utils.rb (hunk 5 above)
Comment (thijsc) [inline lib/appsignal/utils.rb]: Will move it. It's a data object you can pass to the extension, so it's a generic thing by definition. I'm open to suggestions for a better name.

File: lib/appsignal/utils.rb (hunk 1 above)
Comment (thijsc) [inline lib/appsignal/utils.rb]: Looks like this wasn't changed in an earlier pull after all? Everything is rebased on the latest develop.

**Comment by thijsc** [comment]
Improved the specs, ready for review again.

File: spec/lib/appsignal/utils_spec.rb (hunk 4 above)
Comment (thijsc) [inline spec/lib/appsignal/utils_spec.rb]: done

File: spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb (hunk 3 above)
Comment (thijsc) [inline spec/lib/appsignal/integrations/mongo_ruby_driver_spec.rb]: It now tests the content again.

File: lib/appsignal/utils.rb (hunk 1 above)
Comment (tombruijn) [inline lib/appsignal/utils.rb]: No it's on the master branch I think. Better to just undo the change here

File: lib/appsignal/utils.rb [hunk 6]
```diff
@@ -14,12 +14,81 @@ def sanitize(params, only_top_level = false, key_sanitizer = nil)
     end
     extend ClassMethods
 
+    def self.data_generate(body)
+      Data.generate(body)
+    end
+
+    class Data
+      class << self
```
Comment (tombruijn) [inline lib/appsignal/utils.rb]: Is there a reason this has to be a singleton class or can it be a module as well?

File: lib/appsignal/utils.rb (hunk 2 above)
Comment (tombruijn) [inline lib/appsignal/utils.rb]: 👍  can be changed later

File: lib/appsignal/utils.rb (hunk 1 above)
Comment (thijsc) [inline lib/appsignal/utils.rb]: I think we can keep it in. The module construct does nothing currently, it's more clear this way.

File: lib/appsignal/extension.rb [hunk 7]
```diff
@@ -28,5 +28,11 @@ def method_missing(m, *args, &block)
         # Do nothing if the extension methods are not loaded
       end
     end
+
+    class Data
+      def inspect
+        "#<#{self.class.name}:#{self.object_id} #{self.to_s}>"
```
Comment (tombruijn) [inline lib/appsignal/extension.rb]: no need for `self.` with `object_id` and `to_s` methods I think. It is needed for `self.class`

File: spec/lib/appsignal/transaction_spec.rb [hunk 8]
```diff
@@ -450,7 +456,7 @@ def []=(key, val)
           transaction.ext.should_receive(:set_error).with(
             'RSpec::Mocks::Mock',
             '',
-            "[\"line 1\"]"
+            instance_of(Appsignal::Extension::Data)
```
Comment (tombruijn) [inline spec/lib/appsignal/transaction_spec.rb]: only place we don't test the content?

File: lib/appsignal/utils.rb (hunk 6 above)
Comment (thijsc) [inline lib/appsignal/utils.rb]: I think I tried a module but it didn't work in the same way. Not sure. We can refactor this when we move the utils to their own file.

Comment (tombruijn) [inline lib/appsignal/utils.rb]: ok. I'll put it on my list

**Comment by thijsc** [comment]
Processed the last round of feedback.
